File: TypeScript/ExampleTests/common/FandCTyped.ts
import { FandCString } from "./FandCString.js";
import * as _json from "./json.js";

export class FandCTyped {
  f: number;
  c: number;
  notes: string;

  constructor(f: number, c: number, notes: string) {
    this.f = f;
    this.c = c;
    this.notes = notes;
  }

  static fromStringObj(s: FandCString): FandCTyped {
    return new FandCTyped(
      s.f !== "" ? Number(s.f) : 0,
      s.c !== "" ? Number(s.c) : 0,
      s.notes
    );
  }

  toJsonValue(): Record<string, unknown> {
    return {
      f: this.f,
      c: this.c,
      notes: this.notes,
    };
  }

  toJSON(): string { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m: unknown): FandCTyped {
    return new FandCTyped(
      _json.asInt(_json.requireField(m, "f"), "f"),
      _json.asInt(_json.requireField(m, "c"), "c"),
      _json.asString(_json.requireField(m, "notes"), "notes")
    );
  }

  static fromJSON(text: string): FandCTyped {
    return FandCTyped.fromJsonValue(_json.parse(text));
  }

  static toJSONList(items: readonly FandCTyped[]): string {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text: string): FandCTyped[] {
    const raw = _json.asArray(_json.parse(text), "FandCTyped") ?? [];
    return raw.map((e) => FandCTyped.fromJsonValue(e));
  }

  toString(): string {
    return `F=${this.f}, C=${this.c}, Notes=${this.notes}`;
  }

  equals(other: FandCTyped): boolean {
    return this.f === other.f
      && this.c === other.c
      && this.notes === other.notes;
  }
}

File: Python/ExampleTests/common/fand_c_typed.py
from . import json_util as _json
from .fand_c_string import FandCString

class FandCTyped:
    def __init__(self, f: int = 0, c: int = 0, notes: str = ''):
        self.f = f
        self.c = c
        self.notes = notes

    @classmethod
    def from_string_obj(cls, s: FandCString) -> 'FandCTyped':
        return cls(
            int(s.f) if s.f else 0,
            int(s.c) if s.c else 0,
            s.notes
        )

    def to_string_obj(self) -> FandCString:
        return FandCString(
            str(self.f),
            str(self.c),
            str(self.notes)
        )

    @staticmethod
    def to_string_list(items) -> list:
        return [t.to_string_obj() for t in items]

    @staticmethod
    def from_string_list(items) -> list:
        return [FandCTyped.from_string_obj(s) for s in items]

    def to_json_value(self) -> dict:
        return {
            'F': self.f,
            'C': self.c,
            'Notes': self.notes,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'FandCTyped':
        return cls(
            _json.as_int(_json.require(m, 'F'), 'F'),
            _json.as_int(_json.require(m, 'C'), 'C'),
            _json.as_str(_json.require(m, 'Notes'), 'Notes')
        )

    @classmethod
    def from_json(cls, text: str) -> 'FandCTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'FandCTyped')
        return [cls.from_json_value(e) for e in raw]

    def __str__(self):
        return (f'F={self.f}' + ', ' +
                f'C={self.c}' + ', ' +
                f'Notes={self.notes}')

    def _key(self):
        return (self.f, self.c, self.notes)

    def __eq__(self, other):
        if not isinstance(other, FandCTyped):
            return NotImplemented
        return self._key() == other._key()

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'FandCTyped({self})'
